### ISSUE TITLE
Parameter-shift and Adjoint methods don't produce partial derivate vectors in the same layout

### ISSUE THREAD

**Issue by dime10** [PennyLaneAI/catalyst#2450]
Parameter-shift and Adjoint methods don't produce partial derivate vectors in the same layout. As a result, the final Jacobians produced by Catalyst's hybrid differentiation transform are different between the two methods. The transform is currently written for the PS derivative vectors, and produces correct Jacobians for this method.

Example:
```py
import jax.numpy as jnp
import pennylane as qml
from catalyst import *

n_wires = 2
qml.numpy.random.seed(100)
data = qml.numpy.random.rand(n_wires)
weights = qml.numpy.ones([n_wires])

@qml.qnode(qml.device("lightning.qubit", wires=n_wires), diff_method="parameter-shift")
def circuit(weights, data):

    for i in range(n_wires):
        qml.RY(data[i], wires=i)
        qml.RX(weights[i], wires=i)

    return [qml.expval(qml.PauliZ(i)) for i in range(n_wires)]

def loss_fn(weights, data):
    return jnp.array(circuit(weights, data))

@qjit
def grad_fn(weights, data):
    return jacobian(loss_fn, argnums=[0, 1])(weights, data)

results = grad_fn(weights, data)
print(f"Jacobian for arg0:\n{jnp.round(results[0], decimals=4)}")
print(f"Jacobian for arg1:\n{jnp.round(results[1], decimals=4)}")
```
```
Jacobian for arg0:
[[-0.7203 -0.    ]
 [ 0.     -0.8091]]
Jacobian for arg1:
[[-0.2794 -0.    ]
 [ 0.     -0.1485]]
```
which matches results produced by PennyLane. However, with `diff_method="adjoint"`, we get:
```
Jacobian for arg0:
[[ 0.     -0.1485]
 [ 0.     -0.8091]]
Jacobian for arg1:
[[-0.2794 -0.    ]
 [-0.7203  0.    ]]
```

---

Digging deeper, we find a difference in the arrangement of partial derivative vectors produced by the quantum diff methods.
Parameter-shift:
```
[-0.72026,  0]
[-0.279365,  0]

[0,  -0.809078]
[0,  -0.148469]
```
Adjoint:
```
[0,  -0.148469]
[-0.279365,  -0]

[0,  -0.809078]
[-0.72026,  10]
```
These are produced by the [backprop op](https://github.com/PennyLaneAI/catalyst/blob/e53da336915da94480f95971ca0b700a90652054/mlir/include/Gradient/IR/GradientOps.td#L134), which computes vector-Jacobian products for a given cotangent vector. To compute the full Jacobian, we sweep the output dimensions of the function result(s) with one-hot encoded cotangent vectors. The cotangent vectors determine for which result element we compute the gradient vector. Since there are multiple arguments in the function, one backprop call produces one gradient vector for each argument.
So the vectors produced should be as follows:
- cotanget vector [1. 0.]
  - arg0 - res[0]
  - arg1 - res[0]
- cotanget vector [0. 1.]
  - arg0 - res[1]
  - arg1 - res[1]

The vectors are then assembled into full Jacobians for each argument by stacking together gradients for the same argument across the result dimensions.

We can see Parameter-Shift follows this pattern, however Adjoint produces results in a different order:
- arg1 - res[1]
- arg1 - res[0]
- arg0 - res[1]
- arg0 - res[0]

We conclude there are two differences in the Adjoint vectors produced by Lightning that are unaccounted for:
1) the ordering of arguments and the ordering of result elements is reversed
2) the grouping is with respect to arguments rather than with respect to results (i.e. the VJP semantics of the backprop op are not satisfied)

Related to issue pennylaneai/pennylane-lightning#1335 in lightning.

**Comment by dime10** [comment]
Digging further, we can compare the outputs of the parameter-shift and adjoint subroutines directly. Since they only differentiate the quantum circuit, the gradient vectors are computed with respect to gate parameters.
For an example circuit with 8 rotation gates and 4 expval outputs, we get:
```
-- Adjoint --

         g0      g1      g2      g3      g4      g5      g6      g7
        ----------------------------------------------------------------
E[0] | [-0.2794  0.     -0.      0.     -0.     -0.2794 -0.     -0.    ]
E[1] | [-0.7203  0.      0.      0.     -0.     -0.7203 -0.     -0.    ]
E[2] | [ 0.      0.     -0.1485  0.     -0.      0.     -0.     -0.1485]
E[3] | [ 0.     -0.     -0.8091 -0.     -0.     -0.     -0.     -0.8091]

-- Parameter Shift --

         g0      g1      g2      g3      g4      g5      g6      g7
        ----------------------------------------------------------------
E[0] | [-0.2794 -0.7203 -0.      0.      0.      0.      0.     -0.    ]
E[1] | [ 0.     -0.     -0.1485 -0.8091  0.      0.     -0.     -0.    ]
E[2] | [-0.      0.     -0.      0.     -0.2794 -0.7203 -0.     -0.    ]
E[3] | [-0.      0.     -0.      0.      0.      0.     -0.1485 -0.8091]
```

This disagreement must be the source of the problem, since the rest of the differentiation code is the same for both cases.

This is the program used:
```py
n_wires = 4
qml.numpy.random.seed(100)
data = qml.numpy.random.rand(2)
weights = np.ones([3])

@qml.qnode(qml.device("lightning.qubit", wires=n_wires), diff_method="adjoint")
def circuit(weights, data):

    for i in range(n_wires):
        qml.RY(data[i % len(data)], wires=i)
        qml.RX(weights[i % len(weights)], wires=i)

    return [qml.expval(qml.PauliZ(i)) for i in range(n_wires)]
```
If we think about the gradient vector for the first expval (wire 0), only the first two gates (those acting on qubit 0) should affect it, with all other gates having a 0 partial derivative. This is indeed the case for the PS results, but not for the adjoint results. The source of the mix up must come from somewhere within the adjoint method then.
